haskell-auto-import.el

```diff
@@ -0,0 +1,286 @@
+               (imported-items (car (last info 1)))
+               (new-info
+                (append keep-info
+                        (if new-sub-imports-list
+                            (list
+                             (haskell-auto-import-normalize-imported-items
+                              (append imported-items new-sub-imports-list)))
+                          ()
+                            ))))
+          (haskell-auto-import-replace-import module-name new-info))
+      (progn
+        (let ((old-point (point)))
+          (if all-infos
+              (goto-char (+ 1 (car (car all-infos))))
+            (haskell-navigate-imports))
+
+          (insert
+           (haskell-auto-import-print-module-import
+            module-name
+            (list nil nil new-sub-imports-list)))
+          (newline)
+
+          (goto-char old-point))))))
+
+(defun haskell-auto-import-unqualified-all (module-name)
+  "Add a new unqualified import of the provided module. Everything in that
+module is imported."
+  (interactive "MModule Name: ")
+  (haskell-auto-import-unqualified module-name ""))
+
+(defun haskell-auto-import-qualified (module-name qualifier)
+  "Add a new qualified import of the provided module with the provided
+qualifier. Attempts to place the new import next to existing imports of the same
+module if they exist."
+  (interactive "MModule Name: \nMQualifier: ")
+  (let* ((all-infos (haskell-auto-import-module-import-info module-name))
+         (qualified
+          (member-if '(lambda (info)
+                        (car (cdr (cdr info))))
+                     all-infos))
+         (same-qualifier
+          (member-if '(lambda (info)
+                        (string= (car (cdr (cdr info))) qualifier))
+                     all-infos)))
+    (if same-qualifier
+        (message "That module is already imported with the same qualifier. Doing nothing.")
+      (let ((old-point (point)))
+        ;; If the module is already imported in some capacity, put the new
+        ;; import above the existing ones
+        (if all-infos
+            (goto-char (+ 1 (car (car all-infos))))
+          (haskell-navigate-imports))
+
+        (insert
+         (haskell-auto-import-print-module-import
+          module-name
+          (list t qualifier nil)))
+        (newline)
+
+        (goto-char old-point)))))
+
+(defun haskell-auto-import-qualified-and-unqualified (module-name qualifier sub-imports)
+  "Add a new import, both qualified and unqualified. For example, given the
+module `Data.Map', qualifier `Map' and sub-imports `Map', this will add the lines:
+
+import Data.Map (Map)
+import qualified Data.Map as Map"
+
+  (interactive
+   "MModule Name: \nMQualifier: \nMSub-Imports (comma separated, leave blank to import all): ")
+  (haskell-auto-import-qualified module-name qualifier)
+  (haskell-auto-import-unqualified module-name sub-imports))
+
+(defun haskell-auto-import-parse-import-parens (contents)
```

Comment:
In modern elisp style, functions that are part of the package interface are named as `<packagename>-<fname>`, i.e. `haskell-auto-import-parse-import-parens`. Functions that are internal (helper functions and other such functions you don't want clients to depend on) have a double dash after the package name, as `<packagename>--<fname>`. This looks like an internal function, so I'd recommend renaming it to `haskell-auto-import--parse-import-parens`.